### PR TITLE
url safe base64 encoding/decoding added

### PR DESCRIPTION
**Pull request by crea7or**

### CODE REVIEW

**Comment by erankor** [comment]
Hi,

Was just looking for this feature... I see this pull is open for a long time, any chance it will be merged?

Thanks,

Eran

**Comment by doujiang24** [comment]
@rainingmaster will you help to take a look at this PR? thanks.

**Comment by rainingmaster** [comment]
@crea7or Sorry for our delay, could you add some test cases like [set_encode_base64](https://github.com/openresty/set-misc-nginx-module/blob/e29579b95723dd04ecd6b6777cb0d826a7455f42/t/base64.t#L18)?

**Comment by crea7or** [comment]
> @crea7or Sorry for our delay, could you add some test cases like [set_encode_base64](https://github.com/openresty/set-misc-nginx-module/blob/e29579b95723dd04ecd6b6777cb0d826a7455f42/t/base64.t#L18)?

Have no repo right now, so, can I add it right here?
base64url.t

`
# vi:filetype=

use lib 'lib';
use Test::Nginx::Socket;

#repeat_each(3);

plan tests => repeat_each() * 2 * blocks();

no_long_string();

run_tests();

#no_diff();

__DATA__

=== TEST 1: base64url encode
--- config
    location /bar {
        set_encode_base64url $out "?b><d?";
        echo $out;
    }
--- request
    GET /bar
--- response_body
P2I-PGQ_

=== TEST 2: base64url decode
--- config
    location /bar {
        set_decode_base64url $out "P2I-PGQ_";
        echo $out;
    }
--- request
    GET /bar
--- response_body
?b><d?
`

**Comment by crea7or** [comment]
Can I add it right here? Have no repo right now.
[base64url.txt](https://github.com/openresty/set-misc-nginx-module/files/4640904/base64url.txt)

**Comment by rainingmaster** [comment]
> Can I add it right here? Have no repo right now.
> [base64url.txt](https://github.com/openresty/set-misc-nginx-module/files/4640904/base64url.txt)

@crea7or Sorry, I think I don't really understand your mean of "no repo right", I think you can change in [your repo](https://github.com/crea7or/set-misc-nginx-module), and add another commit? If you finish the commit of the test, we can see it performance easier in [here](https://travis-ci.org/github/openresty/set-misc-nginx-module).

By the way, here is [a little change should be fix](https://github.com/openresty/set-misc-nginx-module/pull/45#discussion_r426211283)

**Comment by crea7or** [comment]
@rainingmaster done.

**Comment by rainingmaster** [comment]
@crea7or 
Hi, it seem the style does not fixed yet? https://github.com/openresty/set-misc-nginx-module/pull/45#discussion_r426211283

**Comment by crea7or** [comment]
@rainingmaster spaces fixed

**Comment by erankor** [comment]
ping

**Comment by crea7or** [comment]
why still not merged?

**Comment by erankor** [comment]
Ping!

File: src/ngx_http_set_misc_module.c [hunk 1]
```diff
@@ -55,6 +56,19 @@ static ndk_set_var_t  ngx_http_set_misc_set_decode_base64_filter = {
     NULL
 };
```
Comment (zhuizhuhaomeng) [inline src/ngx_http_set_misc_module.c:58]: two blank lines here

File: src/ngx_http_set_misc_module.c [hunk 2]
```diff
@@ -55,6 +56,19 @@ static ndk_set_var_t  ngx_http_set_misc_set_decode_base64_filter = {
...
 
+static ndk_set_var_t  ngx_http_set_misc_set_encode_base64url_filter = {
+    NDK_SET_VAR_VALUE,
+    (void *) ngx_http_set_misc_set_encode_base64url,
+    1,
+    NULL
+};
+
```
Comment (zhuizhuhaomeng) [inline src/ngx_http_set_misc_module.c:65]: two blank lines here

File: src/ngx_http_set_misc_module.c [hunk 3]
```diff
@@ -55,6 +56,19 @@ static ndk_set_var_t  ngx_http_set_misc_set_decode_base64_filter = {
...
+static ndk_set_var_t  ngx_http_set_misc_set_decode_base64url_filter = {
+    NDK_SET_VAR_VALUE,
+    (void *) ngx_http_set_misc_set_decode_base64url,
+    1,
+    NULL
+};
```
Comment (zhuizhuhaomeng) [inline src/ngx_http_set_misc_module.c:72]: two blank lines here.

**Comment by crea7or** [comment]
done @zhuizhuhaomeng

**Review by zhuizhuhaomeng — APPROVED** [review]

**Pull request merged**